tidy3d/plugins/metrics/README.md

```diff
@@ -0,0 +1,265 @@
+#### Mixing Positional and Keyword Arguments
+
+You can mix positional and keyword arguments when evaluating expressions that contain both unnamed and named variables.
+
+```python
+# Create a mix of named and unnamed variables
+x = Variable(name="x")
+y = Variable()
+expr = x**2 + y**2
+
+result = expr(3, y=4)  # x = 3, y = 4
+
+print(result)  # Outputs: 25
+```
+
+#### A `Metric` is a `Variable`
+
+A `Metric` such as `ModeCoefficient` is a subclass of `Variable`, so all the rules for `Variable` apply to `Metric` as well.
+
+from tidy3d.plugins.metrics import ModeCoefficient, ModePower
+
+```python
+# Define two named metrics
+mode_coeff1 = ModeCoefficient(name="mode_coeff1", monitor_name="monitor1", freqs=[1.0])
+mode_power1 = ModePower(name="mode_power1", monitor_name="monitor2", freqs=[1.0])
+
+# Create an expression using the metrics
+expr = mode_coeff1 + mode_power1
+
+# Assume "data1" and "data2" are SimulationData objects obtained from different simulations
+result = expr(mode_coeff1=data1, mode_power1=data2)
+```
+
+#### Important notes on variable evaluation
+
+- **Unnamed Variables**: If you have an unnamed variable (`Variable()`), you must provide exactly one positional argument during evaluation.
+Providing multiple positional arguments will raise a `ValueError`.
+
+- **Named Variables**: For each named variable, you must provide a corresponding keyword argument.
+If a named variable is missing during evaluation, a `ValueError` will be raised.
+
+- **Multiple Unnamed Variables**: If your expression contains multiple unnamed variables, you cannot provide multiple positional arguments.
+You should assign names to the variables and provide their values as keyword arguments instead.
+
+## Developer notes
+
+### Extending metrics
+
+To implement new metrics, follow these steps:
+
+1. **Subclass the `Metric` base class:**
+
+   Create a new class that inherits from `Metric` and implement the required methods.
+
+   ```python
+   from tidy3d.plugins.metrics import Metric
+
+   class CustomMetric(Metric):
+       monitor_name: str
+
+       def evaluate(self, data: SimulationData) -> NumberType:
+           # Implement custom evaluation logic
+           pass
+   ```
+
+2. **Define attributes and methods:**
+
+   - Add any necessary attributes with type annotations.
+   - Implement the `evaluate` method, which defines how the metric is calculated from simulation data.
+
+3. **Update forward references:**
+
+   If your metrics reference other metrics or expressions, ensure you handle forward references appropriately.
+
+### Extending operators
+
+To extend the `metrics` module with additional operators:
+
+1. **Create a new operator class:**
+
+   Subclass `UnaryOperator` or `BinaryOperator` depending on the operator's arity.
+
+   ```python
+   from tidy3d.plugins.metrics import BinaryOperator
+
+   class CustomOperator(BinaryOperator):
+       _symbol = "??"  # Replace with the operator symbol
+       _format = "({left} {symbol} {right})"
```

Comment:
are these `left` `symbol` `right` keywords? what are the keywords for `UnitaryOperator`?

Reply:
symbol and format are private and ignored by pydantic. they are only used in the `__repr__`. keyword arguments for operators are `operand` for the `UnaryOperator` and `left` and `right` for the `BinaryOperator`. when you derive from either of those, you only need to provide an `evaluate()` implementation that uses those. and if you want your custom operator to be printed correctly, you need to provide `_symbol` and `_format`. although `_format` is actually usually not needed because those should be pretty much the same across all unary and binary operators. so maybe that part is a bit misleading in the readme.